# **Setup / Initial**

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
#pd.set_option('display.width', 500)

df = pd.read_csv('dataset/nba_logreg.csv')
df.head()

name  gp   min  pts  fgm  fga    fg  3p_made  3pa    3p  ftm  \
0   Brandon Ingram  36  27.4  7.4  2.6  7.6  34.7      0.5  2.1  25.0  1.6   
1  Andrew Harrison  35  26.9  7.2  2.0  6.7  29.6      0.7  2.8  23.5  2.6   
2   JaKarr Sampson  74  15.3  5.2  2.0  4.7  42.2      0.4  1.7  24.4  0.9   
3      Malik Sealy  58  11.6  5.7  2.3  5.5  42.6      0.1  0.5  22.6  0.9   
4      Matt Geiger  48  11.5  4.5  1.6  3.0  52.4      0.0  0.1   0.0  1.3   

   fta    ft  oreb  dreb  reb  ast  stl  blk  tov  target_5yrs  
0  2.3  69.9   0.7   3.4  4.1  1.9  0.4  0.4  1.3          0.0  
1  3.4  76.5   0.5   2.0  2.4  3.7  1.1  0.5  1.6          0.0  
2  1.3  67.0   0.5   1.7  2.2  1.0  0.5  0.3  1.0          0.0  
3  1.3  68.9   1.0   0.9  1.9  0.8  0.6  0.1  1.0          1.0  
4  1.9  67.4   1.0   1.5  2.5  0.3  0.3  0.4  0.8          1.0

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1340 entries, 0 to 1339
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         1340 non-null   object 
 1   gp           1340 non-null   int64  
 2   min          1340 non-null   float64
 3   pts          1340 non-null   float64
 4   fgm          1340 non-null   float64
 5   fga          1340 non-null   float64
 6   fg           1340 non-null   float64
 7   3p_made      1340 non-null   float64
 8   3pa          1340 non-null   float64
 9   3p           1329 non-null   float64
 10  ftm          1340 non-null   float64
 11  fta          1340 non-null   float64
 12  ft           1340 non-null   float64
 13  oreb         1340 non-null   float64
 14  dreb         1340 non-null   float64
 15  reb          1340 non-null   float64
 16  ast          1340 non-null   float64
 17  stl          1340 non-null   float64
 18  blk          1340 non-null   float64
 19  tov   

column '3p' has 1329 non-null. That is 11 less than normal. Let's see why that is.

In [5]:
null_3p = df[df['3p'].isnull()]
display(null_3p)

name  gp   min   pts  fgm  fga    fg  3p_made  3pa  3p  ftm  \
338     Ken Johnson  64  12.7   4.1  1.8  3.3  52.8      0.0  0.0 NaN  0.6   
339     Ken Johnson  64  12.7   4.1  1.8  3.3  52.8      0.0  0.0 NaN  0.6   
340   Pete Williams  53  10.8   2.8  1.3  2.1  60.4      0.0  0.0 NaN  0.3   
358   Melvin Turpin  79  24.7  10.6  4.6  9.0  51.1      0.0  0.0 NaN  1.4   
386    Jim Petersen  60  11.9   3.2  1.2  2.4  48.6      0.0  0.0 NaN  0.8   
397   Tom Scheffler  39   6.9   1.3  0.5  1.3  41.2      0.0  0.0 NaN  0.3   
507    Sam Williams  59  18.2   6.1  2.6  4.7  55.6      0.0  0.0 NaN  0.8   
509   Kurt Nimphius  63  17.2   5.3  2.2  4.7  46.1      0.0  0.0 NaN  1.0   
510  Pete Verhoeven  71  17.0   4.9  2.1  4.2  50.3      0.0  0.0 NaN  0.7   
521       Jim Smith  72  11.9   2.9  1.2  2.3  50.9      0.0  0.0 NaN  0.5   
559    Jeff Wilkins  56  18.9   4.7  2.1  4.6  45.0      0.0  0.0 NaN  0.5   

     fta    ft  oreb  dreb  reb  ast  stl  blk  tov  target_5yrs  
338  1.3  43.5   1.4   2.4  3.8  0.3  0.2  0.3  0.9          0.0  
339  1.3  43.5   1.4   2.4  3.8  0.3  0.2  0.3  0.9          0.0  
340  0.8  42.5   0.9   1.9  2.8  0.3  0.4  0.4  0.4          0.0  
358  1.8  78.4   2.0   3.8  5.7  0.5  0.5  1.1  1.5          1.0  
386  1.1  75.8   0.7   1.7  2.5  0.5  0.2  0.5  1.2          1.0  
397  0.5  50.0   0.5   1.5  1.9  0.3  0.2  0.3  0.4          0.0  
507  1.5  55.1   1.5   3.7  5.2  0.6  0.8  1.3  1.1          0.0  
509  1.7  58.3   1.5   3.2  4.7  1.0  0.3  1.3  0.9          1.0  
510  1.0  70.8   1.5   2.1  3.6  0.7  0.6  0.3  0.8          1.0  
521  1.2  45.9   1.0   1.5  2.5  0.6  0.3  0.7  0.7          0.0  
559  0.7  67.5   1.1   3.8  4.9  0.7  0.6  0.8  1.1          1.0

Null values in the '3p' (3-point percentage) column means the player has **No 3-pt attempts:** ('3pa' = 0). This is true for all 11 instance in this case.

Will change it to be 0 instead of null.

In [6]:
df['3p'] = df['3p'].fillna(0)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1340 entries, 0 to 1339
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         1340 non-null   object 
 1   gp           1340 non-null   int64  
 2   min          1340 non-null   float64
 3   pts          1340 non-null   float64
 4   fgm          1340 non-null   float64
 5   fga          1340 non-null   float64
 6   fg           1340 non-null   float64
 7   3p_made      1340 non-null   float64
 8   3pa          1340 non-null   float64
 9   3p           1340 non-null   float64
 10  ftm          1340 non-null   float64
 11  fta          1340 non-null   float64
 12  ft           1340 non-null   float64
 13  oreb         1340 non-null   float64
 14  dreb         1340 non-null   float64
 15  reb          1340 non-null   float64
 16  ast          1340 non-null   float64
 17  stl          1340 non-null   float64
 18  blk          1340 non-null   float64
 19  tov   

In [9]:
df.duplicated().sum()

np.int64(12)

In [11]:
duplicate_rows = df[df.duplicated(keep=False)]

if not duplicate_rows.empty:
    display(duplicate_rows.sort_values(by=list(df.columns)))


Number of duplicated rows (including all instances): 24
Duplicated Rows (all instances):


name  gp   min   pts  fgm   fga    fg  3p_made  3pa    3p  \
368      Charles Jones  29  16.4   3.7  1.3   4.2  31.7      0.7  2.1  31.1   
369      Charles Jones  29  16.4   3.7  1.3   4.2  31.7      0.7  2.1  31.1   
365      Charles Jones  78  20.1   8.4  3.0   5.8  52.0      0.0  0.1   0.0   
366      Charles Jones  78  20.1   8.4  3.0   5.8  52.0      0.0  0.1   0.0   
168      Charles Smith  34   8.6   3.5  1.4   3.7  39.2      0.4  1.4  31.9   
169      Charles Smith  34   8.6   3.5  1.4   3.7  39.2      0.4  1.4  31.9   
162      Charles Smith  60   8.7   2.9  1.0   2.2  44.4      0.0  0.1   0.0   
163      Charles Smith  60   8.7   2.9  1.0   2.2  44.4      0.0  0.1   0.0   
165      Charles Smith  71  30.4  16.3  6.1  12.4  49.5      0.0  0.0   0.0   
166      Charles Smith  71  30.4  16.3  6.1  12.4  49.5      0.0  0.0   0.0   
1334       Chris Smith  80  15.8   4.3  1.6   3.6  43.3      0.0  0.2  14.3   
1335       Chris Smith  80  15.8   4.3  1.6   3.6  43.3      0.0  0.2  14.3   
503      Eddie Johnson  74  20.5   9.3  4.0   8.7  45.9      0.0  0.1   9.1   
504      Eddie Johnson  74  20.5   9.3  4.0   8.7  45.9      0.0  0.1   9.1   
732   Gerald Henderson  43   8.3   2.6  0.9   2.4  35.6      0.1  0.4  21.1   
733   Gerald Henderson  43   8.3   2.6  0.9   2.4  35.6      0.1  0.4  21.1   
338        Ken Johnson  64  12.7   4.1  1.8   3.3  52.8      0.0  0.0   0.0   
339        Ken Johnson  64  12.7   4.1  1.8   3.3  52.8      0.0  0.0   0.0   
823    Marcus Williams  79  16.6   6.8  2.6   6.7  39.5      0.6  2.1  28.2   
824    Marcus Williams  79  16.6   6.8  2.6   6.7  39.5      0.6  2.1  28.2   
970      Mike Dunleavy  82  15.9   5.7  2.0   5.1  40.3      0.6  1.8  34.7   
971      Mike Dunleavy  82  15.9   5.7  2.0   5.1  40.3      0.6  1.8  34.7   
242    Reggie Williams  35  24.5  10.4  4.3  12.2  35.6      0.4  1.7  22.4   
243    Reggie Williams  35  24.5  10.4  4.3  12.2  35.6      0.4  1.7  22.4   

      ftm  fta    ft  oreb  dreb  reb  ast  stl  blk  tov  target_5yrs  
368   0.4  0.8  50.0   0.3   1.1  1.4  1.4  0.6  0.2  1.0          0.0  
369   0.4  0.8  50.0   0.3   1.1  1.4  1.4  0.6  0.2  1.0          0.0  
365   2.3  3.6  64.8   1.8   3.3  5.1  1.6  0.6  0.8  1.8          0.0  
366   2.3  3.6  64.8   1.8   3.3  5.1  1.6  0.6  0.8  1.8          0.0  
168   0.2  0.3  54.5   0.4   0.4  0.8  0.6  0.3  0.2  0.8          1.0  
169   0.2  0.3  54.5   0.4   0.4  0.8  0.6  0.3  0.2  0.8          1.0  
162   0.9  1.3  69.7   0.2   0.9  1.2  1.7  0.6  0.1  0.6          1.0  
163   0.9  1.3  69.7   0.2   0.9  1.2  1.7  0.6  0.1  0.6          1.0  
165   4.0  5.5  72.5   2.4   4.1  6.5  1.5  1.0  1.3  2.1          1.0  
166   4.0  5.5  72.5   2.4   4.1  6.5  1.5  1.0  1.3  2.1          1.0  
1334  1.2  1.5  79.2   0.4   0.8  1.2  2.5  0.6  0.2  0.8          0.0  
1335  1.2  1.5  79.2   0.4   0.8  1.2  2.5  0.6  0.2  0.8          0.0  
503   1.3  2.0  66.4   1.7   2.6  4.4  1.5  0.7  0.2  1.3          1.0  
504   1.3  2.0  66.4   1.7   2.6  4.4  1.5  0.7  0.2  1.3          1.0  
732   0.8  1.1  74.5   0.3   0.9  1.3  0.3  0.2  0.2  0.3          1.0  
733   0.8  1.1  74.5   0.3   0.9  1.3  0.3  0.2  0.2  0.3          1.0  
338   0.6  1.3  43.5   1.4   2.4  3.8  0.3  0.2  0.3  0.9          0.0  
339   0.6  1.3  43.5   1.4   2.4  3.8  0.3  0.2  0.3  0.9          0.0  
823   0.9  1.1  84.7   0.4   1.7  2.1  3.3  0.4  0.0  1.8          0.0  
824   0.9  1.1  84.7   0.4   1.7  2.1  3.3  0.4  0.0  1.8          0.0  
970   1.0  1.2  78.0   0.8   1.8  2.6  1.3  0.7  0.2  1.1          1.0  
971   1.0  1.2  78.0   0.8   1.8  2.6  1.3  0.7  0.2  1.1          1.0  
242   1.4  1.9  72.7   1.6   1.8  3.4  1.7  0.8  0.6  1.8          1.0  
243   1.4  1.9  72.7   1.6   1.8  3.4  1.7  0.8  0.6  1.8          1.0

Need to drop duplicated rows. Some names like Charles Smith have 3 different stat variations. Looking into this, this information is correct. There are 3 different Charles Smith that appeared at different times. Same goes with Charles Jones.

In [12]:
print(f"Number of rows before: {len(df)}")

Number of rows before: 1340


In [13]:
df.drop_duplicates(inplace=True)
print(f"Number of rows after dropping duplicates: {len(df)}")
df.info()

Number of rows after dropping duplicates: 1328
<class 'pandas.core.frame.DataFrame'>
Index: 1328 entries, 0 to 1339
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         1328 non-null   object 
 1   gp           1328 non-null   int64  
 2   min          1328 non-null   float64
 3   pts          1328 non-null   float64
 4   fgm          1328 non-null   float64
 5   fga          1328 non-null   float64
 6   fg           1328 non-null   float64
 7   3p_made      1328 non-null   float64
 8   3pa          1328 non-null   float64
 9   3p           1328 non-null   float64
 10  ftm          1328 non-null   float64
 11  fta          1328 non-null   float64
 12  ft           1328 non-null   float64
 13  oreb         1328 non-null   float64
 14  dreb         1328 non-null   float64
 15  reb          1328 non-null   float64
 16  ast          1328 non-null   float64
 17  stl          1328 non-null   float64
 18  blk   

In [14]:
df.describe()

gp          min          pts          fgm          fga  \
count  1328.000000  1328.000000  1328.000000  1328.000000  1328.000000   
mean     60.411145    17.634413     6.804217     2.630271     5.885542   
std      17.416688     8.322560     4.361669     1.684918     3.595394   
min      11.000000     3.100000     0.700000     0.300000     0.800000   
25%      47.000000    10.875000     3.700000     1.400000     3.300000   
50%      63.000000    16.100000     5.550000     2.100000     4.800000   
75%      77.000000    22.900000     8.800000     3.400000     7.500000   
max      82.000000    40.900000    28.200000    10.200000    19.800000   

                fg      3p_made         3pa           3p          ftm  \
count  1328.000000  1328.000000  1328.00000  1328.000000  1328.000000   
mean     44.184639     0.247741     0.77869    19.177485     1.298117   
std       6.132759     0.384548     1.06351    16.071943     0.987329   
min      23.800000     0.000000     0.00000     0.000000     0.000000   
25%      40.200000     0.000000     0.00000     0.000000     0.600000   
50%      44.100000     0.100000     0.30000    22.200000     1.000000   
75%      47.900000     0.400000     1.12500    32.500000     1.600000   
max      73.700000     2.300000     6.50000   100.000000     7.700000   

               fta           ft         oreb         dreb          reb  \
count  1328.000000  1328.000000  1328.000000  1328.000000  1328.000000   
mean      1.822139    70.325226     1.009714     2.027636     3.036446   
std       1.322653    10.562158     0.777551     1.362236     2.060241   
min       0.000000     0.000000     0.000000     0.200000     0.300000   
25%       0.900000    64.700000     0.400000     1.000000     1.500000   
50%       1.500000    71.250000     0.800000     1.700000     2.500000   
75%       2.300000    77.600000     1.400000     2.600000     4.000000   
max      10.200000   100.000000     5.300000     9.600000    13.900000   

               ast          stl          blk          tov  target_5yrs  
count  1328.000000  1328.000000  1328.000000  1328.000000  1328.000000  
mean      1.551205     0.619051     0.368675     1.193599     0.620482  
std       1.475731     0.410972     0.429680     0.723966     0.485450  
min       0.000000     0.000000     0.000000     0.100000     0.000000  
25%       0.600000     0.300000     0.100000     0.700000     0.000000  
50%       1.100000     0.500000     0.200000     1.000000     1.000000  
75%       2.000000     0.800000     0.500000     1.500000     1.000000  
max      10.600000     2.500000     3.900000     4.400000     1.000000

Output new, cleaned CSV file

In [15]:
import os

output_dir = 'dataset'
if not os.path.exists(output_dir):
    os.makedirs(output_dir)

output_filepath = os.path.join(output_dir, 'nba_cleaned.csv')
df.to_csv(output_filepath, index=False)
print(f"Cleaned data saved to {output_filepath}")

Cleaned data saved to dataset/nba_cleaned.csv
